src/std_extensions/quantum.rs

```diff
@@ -1,19 +1,168 @@
+// SAFETY: unsafe block should be ok as the value is definitely not zero.
+#[allow(clippy::assertions_on_constants)]
+pub const PRECISION_TYPE_PARAM: TypeParam = TypeParam::bounded_nat(unsafe {
+    assert!(PRECISION_BOUND > 0);
+    NonZeroU64::new_unchecked(PRECISION_BOUND as u64)
+});
+
+/// Get the precision of the specified type argument or error if the argument is invalid.
+pub(super) fn get_precision(arg: &TypeArg) -> Result<u8, TypeArgError> {
+    match arg {
+        TypeArg::BoundedNat { n } if is_valid_precision(*n as u8) => Ok(*n as u8),
+        _ => Err(TypeArgError::TypeMismatch {
+            arg: arg.clone(),
+            param: PRECISION_TYPE_PARAM,
+        }),
+    }
+}
+
+pub(super) const fn type_arg(precision: u8) -> TypeArg {
+    TypeArg::BoundedNat {
+        n: precision as u64,
+    }
+}
+
+/// An angle
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct ConstAngle {
+    precision: u8,
+    value: u64,
+}
+
+impl ConstAngle {
+    /// Create a new [`ConstAngle`]
+    pub fn new(precision: u8, value: u64) -> Result<Self, ConstTypeError> {
+        if !is_valid_precision(precision) {
+            return Err(ConstTypeError::CustomCheckFail(
+                crate::types::CustomCheckFailure::Message("Invalid angle precision.".to_owned()),
+            ));
+        }
+        if value >= (1u64 << precision) {
+            return Err(ConstTypeError::CustomCheckFail(
+                crate::types::CustomCheckFailure::Message(
+                    "Invalid unsigned integer value.".to_owned(),
+                ),
+            ));
+        }
+        Ok(Self { precision, value })
+    }
+
+    /// Returns the value of the constant
+    pub fn value(&self) -> u64 {
+        self.value
+    }
+
+    /// Returns the precision of the constant
+    pub fn precision(&self) -> u8 {
+        self.precision
+    }
+}
+
+#[typetag::serde]
+impl CustomConst for ConstAngle {
+    fn name(&self) -> SmolStr {
+        format!("a(2π*{}/{})", self.value, 1u64 << self.precision).into()
+    }
+    fn check_custom_type(&self, typ: &CustomType) -> Result<(), CustomCheckFailure> {
+        if typ.clone() == angle_custom_type(type_arg(self.precision)) {
+            Ok(())
+        } else {
+            Err(CustomCheckFailure::Message(
+                "Angle constant type mismatch.".into(),
+            ))
+        }
+    }
+    fn equal_consts(&self, other: &dyn CustomConst) -> bool {
+        crate::values::downcast_equal_consts(self, other)
+    }
+}
+
+fn atrunc_sig(arg_values: &[TypeArg]) -> Result<FunctionType, SignatureError> {
+    let [arg0, arg1] = collect_array(arg_values);
+    let m: u8 = get_precision(arg0)?;
+    let n: u8 = get_precision(arg1)?;
+    if m < n {
+        return Err(SignatureError::InvalidTypeArgs);
+    }
+    Ok(FunctionType::new(
+        vec![angle_type(arg0.clone())],
+        vec![Type::new_sum(vec![angle_type(arg1.clone()), ERROR_TYPE])],
```

Comment:
what causes the error state? It's not in the spec

Reply:
Right, this can't error -- I think I was confusing this with an operation that would fail (instead of rounding down) if the angles didn't match exactly. That could actually be a useful operation, and I should maybe add it -- or replace `awiden<M,N>` with a more general one that doesn't require M <= N but can fail if M > N and the angle doesn't "fit" exactly.